clients/src/main/java/org/apache/kafka/clients/producer/internals/Sender.java

```diff
@@ -251,13 +252,15 @@ private void handleProduceResponse(ClientResponse response, Map<TopicPartition,
             // if we have a response, parse it
             if (response.hasResponse()) {
                 ProduceResponse produceResponse = new ProduceResponse(response.responseBody());
-                for (Map.Entry<TopicPartition, ProduceResponse.PartitionResponse> entry : produceResponse.responses()
-                                                                                                         .entrySet()) {
+                for (Map.Entry<TopicPartition, ProduceResponse.PartitionResponse> entry : produceResponse.responses().entrySet()) {
                     TopicPartition tp = entry.getKey();
                     ProduceResponse.PartitionResponse partResp = entry.getValue();
                     Errors error = Errors.forCode(partResp.errorCode);
                     RecordBatch batch = batches.get(tp);
-                    completeBatch(batch, error, partResp.baseOffset, correlationId, now);
+                    completeBatch(batch, error, partResp.baseOffset, partResp.timestamp, correlationId, now);
+                    if (error == Errors.NONE)
+                        metadata.recordTopicTimestampType(tp.topic(),
```

Comment:
I understand that producer learns about the type of the timestamp after it gets first successful produce response. We are using the type of the timestamp in KafkaProducer.send() to set timestamp in ProducerRecord. I see that we default to handling timestamp as CreateTime timestamp if we don't know the type yet. What is the impact of doing this if the correct type turns out to be LogAppendTIme?  Why don't we just always set timestamp on Producer as if is CreateTime (basically, set local producer time if timestamp in ProducerRecord is null), and broker then overwrites it if the timestamp type is LogAppendTime. Otherwise, looks like lots of added complexity just to decide whether set timestamp on the producer or not.